src/utils/cuda/cuda_mem_component.c

```diff
@@ -27,16 +29,59 @@ static xccl_status_t xccl_cuda_mem_free(void *ptr)
     return XCCL_OK;
 }
 
+static xccl_status_t xccl_cuda_alloc_resources()
+{
+    int i;
+
+    CUDACHECK(cudaStreamCreateWithFlags(&xccl_cuda_mem_component.stream,
+                                        cudaStreamNonBlocking));
+    CUDACHECK(cudaHostAlloc((void**)&xccl_cuda_mem_component.stream_requests,
+              NUM_STREAM_REQUESTS * sizeof(xccl_cuda_mem_component_stream_request_t),
+              cudaHostAllocMapped));
+    for (i = 0; i < NUM_STREAM_REQUESTS; i++) {
+        xccl_cuda_mem_component.stream_requests[i].is_free = 1;
```

Comment:
this looks like a pool of stream_reqs - is it thread safe (does not look like ) ? Can we use ucs mpool for this ?

Reply:
No, it's not thread safe, we need some global lock to access it. And I like the idea of using mpool, should work here, but let's make it in next PR?

Reply:
sure 👍